### PR TITLE
planner: refine function `accessPathsForConds` to avoid wrong indexmerge plan

### LINKED ISSUE #16837: wrong index merge plan for unique index
## Development Task

```
mysql> create table t(a int, b int , c int, d int, e int, unique key idx_ab(a,b), unique key(c), unique key(d));

mysql> explain select /*+ use_index_merge(t,c,idx_ab) */ * from t where a = 1 or (e = 1 and c = 1);
+-------------------------------+---------+-----------+-----------------------------+---------------------------------------------+
| id                            | estRows | task      | access object               | operator info                               |
+-------------------------------+---------+-----------+-----------------------------+---------------------------------------------+
| IndexMerge_8                  | 11.00   | root      |                             |                                             |
| ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t, index:idx_ab(a, b) | range:[1,1], keep order:false, stats:pseudo |
| ├─IndexRangeScan_6(Build)     | 1.00    | cop[tikv] | table:t, index:c(c)         | range:[1,1], keep order:false, stats:pseudo |
| └─TableRowIDScan_7(Probe)     | 11.00   | cop[tikv] | table:t                     | keep order:false, stats:pseudo              |
+-------------------------------+---------+-----------+-----------------------------+---------------------------------------------+
```

Filter `e = 1` is lost in the plan.

### PR DESCRIPTION
**Pull request by lawyerphx**
### What problem does this PR solve?

Issue Number: close #16837 

Problem Summary:

### What is changed and how it works?

What's Changed:
* Add function `ConvertToIndexMergePath` for to-do works.
* Continue when we have no `Accessconds` and some `TableFilter` rules.

How it Works:
* Avoid pushing table filter to the merged indexes.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects
- Performance regression
    - Consumes more CPU

### Release note 
* Fixed the bug that wrong index merge plans are generated for unique index.

### CODE REVIEW

File: planner/core/stats.go [hunk 1]
```diff
@@ -363,6 +363,15 @@ func (ds *DataSource) isInIndexMergeHints(name string) bool {
 	return false
 }
 
+// If AccessConds is empty or tableFilter is not empty, we ignore the access path.
+// Now these conditions are too strict.
+// For example, a sql `select * from t where a > 1 or (b < 2 and c > 3)` and table `t` with indexes
+// on a and b separately. we can generate a `IndexMergePath` with table filter `a > 1 or (b < 2 and c > 3)`.
+// TODO: solve the above case
+func (ds *DataSource) ConvertToIndexMergePath(path *util.AccessPath) bool {
```
Comment (eurekaka) [inline planner/core/stats.go:371]: This function and its name are pretty confusing, why do we need this?

Comment (lawyerphx) [reply planner/core/stats.go:371]: OK.Because there are two places (line 390 and  line 416) that need to add

Comment (lawyerphx) [reply planner/core/stats.go:371]: Do you have any suggestions about the format?

Comment (eurekaka) [reply planner/core/stats.go:367]: I think just adding this comment in one of the check places is fine, we don't need this empty function actually.

Comment (lawyerphx) [reply planner/core/stats.go]: @eurekaka empty function removed now.

**Comment by lawyerphx** [comment]
PTAL ~ @winoros

File: planner/core/integration_test.go [hunk 2]
```diff
@@ -906,3 +906,17 @@ func (s *testIntegrationSuite) TestIssue16290And16292(c *C) {
 		tk.MustQuery("select count(distinct b) from (select * from t ta union all select * from t tb) t;").Check(testkit.Rows("1"))
 	}
 }
+
+func (s *testIntegrationSerialSuite) TestIssue16837(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a int,b int,c int,d int,e int,unique key idx_ab(a,b),unique key(c),unique key(d))")
+	tk.MustQuery("explain select /*+ use_index_merge(t,c,idx_ab) */ * from t where a = 1 or (e = 1 and c = 1)").Check(testkit.Rows(
```
Comment (zz-jason) [inline planner/core/integration_test.go:915]: can we optimize it to use index merge?

Comment (lawyerphx) [reply planner/core/integration_test.go:915]: It seems to me that adding a filter on `IndexRangeScan` can solve this prob. @zz-jason

Comment (zz-jason) [reply planner/core/integration_test.go:915]: exactly

Comment (lawyerphx) [reply planner/core/integration_test.go:915]: Currently, we are using this mechanism in `from (a>1 and b>1) or c=1 ` with `key(a,b), key(c)`

Comment (lawyerphx) [reply planner/core/integration_test.go:915]: I will try to solve this later. (after labor day holiday)

**Pull request merged**